recipes/binutils/all/conanfile.py

```diff
@@ -1,70 +1,512 @@
+
+    def test_freebsd_x86_64(self):
+        self._test_osarch_to_gnutriplet(_ArchOs(arch="x86_64", os="FreeBSD"), _GNUTriplet(machine="x86_64", vendor="pc", os="freebsd", abi=None), "x86_64-pc-freebsd")
+
+    def test_baremetal_x86(self):
+        self._test_osarch_to_gnutriplet(_ArchOs(arch="x86", os="baremetal"), _GNUTriplet(machine="i686", vendor=None, os=None, abi="elf"), "i686-elf")
+
+    def test_baremetal_x86_64(self):
+        archos = _ArchOs(arch="x86_64", os="baremetal")
+        self._test_osarch_to_gnutriplet(archos, _GNUTriplet(machine="x86_64", vendor=None, os=None, abi="elf"), "x86_64-elf")
+        self.assertTrue(archos.is_compatible(_GNUTriplet.from_text("x86_64-elf")))
+        self.assertTrue(archos.is_compatible(_GNUTriplet.from_text("x86_64-none-elf")))
+        self.assertTrue(archos.is_compatible(_GNUTriplet.from_text("x86_64-unknown-elf")))
+
+    def test_baremetal_armv7(self):
+        archos = _ArchOs(arch="armv7", os="baremetal")
+        self._test_osarch_to_gnutriplet(archos, _GNUTriplet(machine="arm", vendor=None, os="none", abi="eabi"), "arm-none-eabi")
+        self.assertTrue(archos.is_compatible(_GNUTriplet.from_text("arm-none-eabi")))
+
+    def test_baremetal_armv8(self):
+        self._test_osarch_to_gnutriplet(_ArchOs(arch="armv8", os="baremetal"), _GNUTriplet(machine="aarch64", vendor=None, os="none", abi="elf"), "aarch64-none-elf")
+
+    def test_baremetal_riscv32(self):
+        self._test_osarch_to_gnutriplet(_ArchOs(arch="riscv32", os="baremetal"), _GNUTriplet(machine="riscv32", vendor=None, os="unknown", abi="elf"), "riscv32-unknown-elf")
+
+    def test_baremetal_riscv64(self):
+        self._test_osarch_to_gnutriplet(_ArchOs(arch="riscv64", os="baremetal"), _GNUTriplet(machine="riscv64", vendor=None, os="unknown", abi="elf"), "riscv64-unknown-elf")
+
+    def test_android_armv7(self):
+        self._test_osarch_to_gnutriplet(_ArchOs(arch="armv7", os="Android", extra={"os.api_level": "31"}), _GNUTriplet(machine="arm", vendor=None, os="linux", abi="androideabi31"), "arm-linux-androideabi31")
+
+    def test_android_armv8(self):
+        self._test_osarch_to_gnutriplet(_ArchOs(arch="armv8", os="Android", extra={"os.api_level": "24"}), _GNUTriplet(machine="aarch64", vendor=None, os="linux", abi="android24"), "aarch64-linux-android24")
+
+    def test_android_x86(self):
+        self._test_osarch_to_gnutriplet(_ArchOs(arch="x86", os="Android", extra={"os.api_level": "16"}), _GNUTriplet(machine="i686", vendor=None, os="linux", abi="android16"), "i686-linux-android16")
+
+    def test_android_x86_64(self):
+        self._test_osarch_to_gnutriplet(_ArchOs(arch="x86_64", os="Android", extra={"os.api_level": "29"}), _GNUTriplet(machine="x86_64", vendor=None, os="linux", abi="android29"), "x86_64-linux-android29")
+        self.assertEqual(_ArchOs(arch="x86_64", os="Android", extra={"os.api_level": "25"}), _ArchOs.from_triplet(_GNUTriplet.from_text("x86_64-linux-android29")))
```

Comment:
Here, I test the correct conversion between an android conan os/arch and gnu triplet.

Reply:
I copied these triplets from the latest android ndk archive.